primitives/src/crh/poseidon/batched_crh.rs

```diff
@@ -0,0 +1,537 @@
+                matrix_mix_short::<F,P>(&mut vec_state[i]);
+            }
+        }
+    }
+}
+
+
+impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonParameters<Fr = F>> BatchFieldBasedHash for PoseidonBatchHash<F, P> {
+    type Data = F;
+    type BaseHash = PoseidonHash<F, P>;
+
+    fn batch_evaluate(input_array: &[F]) -> Result<Vec<F>, Error> {
+
+        // Input:
+        // This function calculates the hashes of inputs by groups of the rate P::R.
+        // The inputs are arranged in an array and arranged as consecutive chunks
+        // Example:
+        // For P::R = 2,
+        // (d_00, d_01, d_10, d_11, d_20, d_21, ...
+        // where d_00 and d_01 are the inputs to the first hash
+        // d_10 and d_11 are the inputs to the second hash ... etc..
+        // Output:
+        // The output is returned as an array of size input length / P::R
+
+        use rayon::prelude::*;
+
+        // Checks that size of input/output vector
+        assert_eq!(P::T - P::R, 1, "The assumption that the capacity is one field element is not satisfied.");
+        let array_length = input_array.len() / P::R;
+        assert_eq!(input_array.len() % P::R, 0, "The length of the input data array is not a multiple of the rate.");
+        assert_ne!(input_array.len(), 0, "Input data array does not contain any data.");
+
+        // Assign pre-computed values of the state vector equivalent to a permutation with zero element state vector
+        let mut state_z = Vec::new();
+        for i in 0..P::T {
+            state_z.push(P::AFTER_ZERO_PERM[i]);
+        }
+
+        // Copy the result of the permutation to a vector of state vectors of the length equal to the length of the input
+        // state is a vector of T-element state vector.
+        let mut state = Vec::new();
+        for _i in 0..array_length {
+            state.push(state_z.clone());
+        }
+
+        // input_idx is to scan the input_array
+        let mut input_idx = 0;
+
+        // Add the input data in chunks of rate size to the state vectors
+        for k in 0..array_length {
+            for j in 0..P::R {
+                state[k][j] += &input_array[input_idx];
+                input_idx += 1;
+            }
+        }
+
+        // Calculate the chunk size to split the state vector
+        let cpus = rayon::current_num_threads();
+        let chunk_size = (array_length as f64/ cpus as f64).ceil() as usize;
+
+        // apply permutation to different chunks in parallel
+        state.par_chunks_mut(chunk_size)
+            .for_each(| p1| {
+                Self::poseidon_perm_gen(p1);
+            });
+
+        // write the result of the hash extracted from the state vector to the output vector
+        let mut output_array=Vec::new();
+        for k in 0..array_length {
+            output_array.push(state[k][0]);
+        }
+        Ok(output_array)
+    }
+
+    fn batch_evaluate_in_place(input_array: &mut[F], output_array: &mut[F]) {
```

Comment:
Is it worth to duplicate almost all code of batch_evaluate() to serve this functionality?

Reply:
It avoids some copies, so I think so